src/common/ad_tx.c

```diff
@@ -0,0 +1,471 @@
+int
+ad_tx_assign(struct ad_tx *tx, void *addr, daos_size_t size, uint32_t val)
+{
+	struct umem_act_item	*it_undo, *it_redo;
+
+	if (addr == NULL || (size != 1 && size != 2 && size != 4))
+		return -DER_INVAL;
+
+	if (tx == NULL) {
+		if (size == 1)
+			*((uint8_t *)addr) = (uint8_t)val;
+		else if (size == 2)
+			*((uint16_t *)addr) = (uint16_t)val;
+		else
+			*((uint32_t *)addr) = val;
+		return 0;
+	}
+
+	D_ALLOC_ACT(it_undo, UMEM_ACT_ASSIGN, size);
+	if (it_undo == NULL)
+		return -DER_NOMEM;
+
+	D_ALLOC_ACT(it_redo, UMEM_ACT_ASSIGN, size);
+	if (it_redo == NULL) {
+		D_FREE(it_undo);
+		return -DER_NOMEM;
+	}
+
+	it_undo->it_act.ac_assign.addr = ad_ptr2addr(addr);
+	it_undo->it_act.ac_assign.size = size;
+	if (size == 1)
+		it_undo->it_act.ac_assign.val = *((uint8_t *)addr);
+	else if (size == 2)
+		it_undo->it_act.ac_assign.val = *((uint16_t *)addr);
+	else
+		it_undo->it_act.ac_assign.val = *((uint32_t *)addr);
+	AD_TX_ACT_ADD(tx, it_undo, ACT_UNDO);
+
+	it_redo->it_act.ac_assign.addr = ad_ptr2addr(addr);
+	it_redo->it_act.ac_assign.size = size;
+	it_undo->it_act.ac_assign.val = val;
+	AD_TX_ACT_ADD(tx, it_redo, ACT_REDO);
+
+	return 0;
+}
+
+/**
+ * memset a storage region, save the operation for redo (and old value for undo if it's
+ * required by @save_old).
+ */
+int
+ad_tx_set(struct ad_tx *tx, void *addr, char c, daos_size_t size, bool save_old)
+{
+	struct umem_act_item	*it_undo, *it_redo;
+
+	if (addr == NULL || size == 0 || size > UMEM_ACT_PAYLOAD_MAX_LEN)
+		return -DER_INVAL;
+
+	if (tx == NULL) {
+		memset(addr, c, size);
+		return 0;
+	}
+
+	D_ALLOC_ACT(it_redo, UMEM_ACT_SET, size);
+	if (it_redo == NULL)
+		return -DER_NOMEM;
+
+	if (save_old) {
+		D_ALLOC_ACT(it_undo, UMEM_ACT_COPY, size);
+		if (it_undo == NULL) {
+			D_FREE(it_redo);
+			return -DER_NOMEM;
+		}
+
+		memcpy(it_undo->it_act.ac_copy.payload, addr, size);
+		it_undo->it_act.ac_copy.addr = ad_ptr2addr(addr);
+		it_undo->it_act.ac_copy.size = size;
+		AD_TX_ACT_ADD(tx, it_undo, ACT_UNDO);
+	}
+
+	it_redo->it_act.ac_set.addr = ad_ptr2addr(addr);
+	it_redo->it_act.ac_set.size = size;
+	it_redo->it_act.ac_set.val = c;
+	AD_TX_ACT_ADD(tx, it_redo, ACT_REDO);
+
+	return 0;
+}
+
+/**
+ * memmove a storage region, save the operation for redo and old memory content for undo.
+ */
+int
+ad_tx_move(struct ad_tx *tx, void *dst, void *src, daos_size_t size)
+{
+	struct umem_act_item	*it_undo, *it_redo;
+
+	if (dst == NULL || src == NULL || size == 0 || size > UMEM_ACT_PAYLOAD_MAX_LEN)
+		return -DER_INVAL;
+
+	if (tx == NULL) {
+		memmove(dst, src, size);
+		return 0;
+	}
+
+	D_ALLOC_ACT(it_undo, UMEM_ACT_COPY, size);
+	if (it_undo == NULL)
+		return -DER_NOMEM;
+
+	D_ALLOC_ACT(it_redo, UMEM_ACT_MOVE, size);
+	if (it_redo == NULL) {
+		D_FREE(it_undo);
+		return -DER_NOMEM;
+	}
+
+	memcpy(it_undo->it_act.ac_copy.payload, dst, size);
+	it_undo->it_act.ac_copy.addr = ad_ptr2addr(dst);
+	it_undo->it_act.ac_copy.size = size;
+	AD_TX_ACT_ADD(tx, it_undo, ACT_UNDO);
```

Comment:
I think we probably should copy the whole region:
 addr = min(dst, src)
 end = max(dst + size, src + size)
 ac_copy.addr = addr;
 ac_copy.size = (end - addr);
 memcpy(....payload, addr, (end -addr))

Reply:
OK, I see your point, if the dst and src buf overlapped, better to protect it by you suggested method.

But if the dst and src buf don't overlap, seems we only need to protect dst buf here? (user can explicitly protect src buf if needed).

Reply:
as discussed, the original code is fine for this.